oak_io/src/handle.rs

```diff
@@ -167,20 +162,20 @@ pub fn extract_handles<T: HandleVisit>(msg: &mut T) -> Vec<Handle> {
 /// assert_eq!(thing, Thing { handle: 42 });
 /// ```
 pub fn inject_handles<T: HandleVisit>(msg: &mut T, handles: &[Handle]) -> Result<(), OakError> {
-    let mut handles = handles.iter();
-    let mut result = Ok(());
-    // TODO(#1599): Remove the `let _` when the underlying issue is fixed.
-    let _ = msg.visit(|handle| {
-        if let Some(to_inject) = handles.next() {
-            *handle = *to_inject;
+    msg.fold(Ok(handles.iter()), |handles, handle| {
+        let mut handles = handles?;
+        let to_inject = handles.next().ok_or(OakError::ProtobufDecodeError(None))?;
+        *handle = *to_inject;
+        Ok(handles)
+    })
+    // Check that there are no remaining handles
+    .and_then(|mut remaining_handles| {
+        if remaining_handles.next().is_some() {
+            Err(OakError::ProtobufDecodeError(None))
```

Comment:
Could you log something meaningful here? As it is, if this happens I expect we will spend hours trying to debug it :)